syntax = "proto3";

package clairvoyant;

//Edge Service interface
message NodeInfo {
  string node_id = 1;
  string node_ip = 2;
  double arrival_time = 3;
  double contact_time = 4;
  //Fill other useful info
  repeated Coordinate contact_points = 5;

}

message Segment {
  string segment_id = 1;
  int32 segment_size = 2;
  string segment_name = 3; 
}

message SegmentOwner{
    Segment segment = 1;
    NodeInfo node_info = 2;
}

//Cloud service interface
message Coordinate {
  double x = 1;
  double y = 2;
  double speed = 3;
  double time = 4;
  double distance = 5;
}

message Route {
  int32 route_id = 1;
  repeated Coordinate points = 2;
}


message VideoRequest {
  string video_id = 1;
  Route route = 2;
}

message VideoReply {
  int64 token_id = 1;
  repeated string urls = 2;
}

message StatusReply{
  string status = 1;
}

message MissedDeliveryRequest {
    int64 token_id = 1;
    repeated Segment segments = 2;
    string video_id = 3;
    string node_id = 4;
}

message DownloadRequest {
    int64 token_id = 1;
    repeated Segment segments = 2;
    map<string, string> segment_sources = 3;
    int64 arrival_time = 4;
    double contact_time = 5;
    repeated Coordinate contact_points = 6;    
}

message DownloadReply {
  int64 token_id = 1;
  repeated string segment_ids = 2;  
}

message DownloadCompleteRequest {
  repeated string segment_ids = 1;
}

message CVRequest{
  oneof request {
    VideoRequest videorequest = 1;
    MissedDeliveryRequest misseddeliveryrequest = 2;
    DownloadCompleteRequest downloadcompleterequest = 3;
  }
}


message CVReply{
  oneof reply{
    VideoReply videoreply = 1;
    StatusReply status = 2;
  }
}

message ModelUpdateRequest {
  string node_id = 1;
  map<int64,double> model=2;
}

message UpdateRequest {
  oneof update{
    ModelUpdateRequest modelupdate = 1;
  }
}

service MonitoringServer {
  rpc HandleUpdateRequest (UpdateRequest) returns (StatusReply);
}

service CVServer {
  rpc HandleCVRequest (CVRequest) returns (CVReply);
    
}

service EdgeServer {
  rpc HandleDownloadRequest(DownloadRequest) returns (DownloadReply);
}

